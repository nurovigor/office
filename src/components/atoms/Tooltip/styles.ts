import { css } from '@linaria/core';
import { boxShadow, colors } from 'src/components/_settings/_variables';

const paddingBottomPosition = '40px 20px';
const paddingTopPosition = '100px';

export const TooltipWrapper = css`
	display: block;
	position: relative;
`;

export const TooltipTip = css`
	position: absolute;
	border-radius: 4px;
	left: 50%;
	transform: translateX(-50%);
	padding: ${paddingBottomPosition};
	color: ${colors.blackColor};
	background: ${colors.whiteSmokeColor};
	font-size: 14px;
	font-family: sans-serif;
	line-height: 1;
	z-index: 100;
	min-width: 200px;
	-webkit-box-shadow: ${boxShadow.shadow};
	-moz-box-shadow: ${boxShadow.shadow};
	box-shadow: ${boxShadow.shadow};

	&::before {
		content: ' ';
		left: 49%;
		top: calc(6px * -1);
		border: 7px solid ${colors.whiteSmokeColor};
		height: 0;
		width: 0;
		position: absolute;
		pointer-events: none;
		margin-left: calc(${paddingBottomPosition} * -1);
		transform: rotate(45deg);
	}

	&.Top {
		top: calc(${paddingTopPosition} * -5);

		&::before {
			top: 98%;
			border-bottom-color: ${colors.whiteSmokeColor};
		}
	}
`;
