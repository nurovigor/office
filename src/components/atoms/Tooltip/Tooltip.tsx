import React, { ReactNode, useState, useRef } from 'react';
import cn from 'classnames';
import { Button } from 'src/components/atoms/Button';
import { useOverflowBottom } from 'src/hooks/useOverflowBottom';
import { useOnClickOutside } from 'src/hooks/useOutsideClick';
import { TooltipTip, TooltipWrapper } from './styles';

type TooltipPropsType = {
	children: ReactNode;
	content: ReactNode | string;
	showModal: () => void;
};

export const Tooltip: React.FC<TooltipPropsType> = ({ children, content, showModal }) => {
	//local states
	const [active, setActive] = useState(false);
	const [isOverflowBottom, setIsOverFlowBottom] = useState<boolean | undefined>(false);
	//refs
	const boxRef = useRef<HTMLDivElement>(null);
	const btnRef = useRef<HTMLButtonElement>(null);

	useOnClickOutside(boxRef, btnRef, setActive, showModal);

	useOverflowBottom(boxRef, setIsOverFlowBottom, [active]);

	const handleTriggerClick = () => {
		setActive(true);
	};

	return (
		<div className={TooltipWrapper}>
			<div data-testid="tooltip-trigger" onClick={handleTriggerClick}>
				{children}
			</div>
			{active && (
				<div
					className={cn(TooltipTip, {
						Top: isOverflowBottom
					})}
					ref={boxRef}
				>
					{content}
					<Button data-testid="show-modal" ref={btnRef} onClick={showModal}>
						Edit
					</Button>
				</div>
			)}
		</div>
	);
};
