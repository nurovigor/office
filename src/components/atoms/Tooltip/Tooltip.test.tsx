import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Tooltip } from 'src/components/atoms/Tooltip/Tooltip';
import userEvent from '@testing-library/user-event';

describe('Tooltip component', () => {
	let callBack = jest.fn();

	it('renders Tooltip', () => {
		const { asFragment } = render(
			<Tooltip content={<h1>Content</h1>} showModal={callBack}>
				<h2>Children</h2>
			</Tooltip>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('renders Tooltip with content', () => {
		const { asFragment } = render(
			<Tooltip content={<h1>Content</h1>} showModal={callBack}>
				<h2>Children</h2>
			</Tooltip>
		);
		userEvent.hover(screen.getByTestId('tooltip-trigger'));
		expect(asFragment()).toMatchSnapshot();
	});

	it('click button Edit works', () => {
		render(
			<Tooltip content={<h1>Content</h1>} showModal={callBack}>
				<h2>Children</h2>
			</Tooltip>
		);
		fireEvent.click(screen.getByTestId('tooltip-trigger'));
		fireEvent.click(screen.getByText(/edit/i));
		expect(callBack).toBeCalled();
	});
});
