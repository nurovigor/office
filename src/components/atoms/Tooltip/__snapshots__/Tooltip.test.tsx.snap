// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component renders Tooltip 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="tooltip-trigger"
    >
      <h2>
        Children
      </h2>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip component renders Tooltip with content 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="tooltip-trigger"
    >
      <h2>
        Children
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
